File: Python3/Sum of Left Leaves.py
#1 DFS
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        if root.left and not root.left.left and not root.left.right:
            return root.left.val + self.sumOfLeftLeaves(root.right)
        return self.sumOfLeftLeaves(root.left) + self.sumOfLeftLeaves(root.right)   # isn't leave



#Another
#def sumOfLeftLeaves(self, root, isLeft=False):
#    if not root: return 0
#    if not (root.left or root.right): return root.val * isLeft
#    return self.sumOfLeftLeaves(root.left, True) + self.sumOfLeftLeaves(root.right)

#2 BFS
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        queue = collections.deque([root])
        sum = 0
        while queue:
            k = len(queue)
            for i in range(k):
                cur = queue.popleft()
                if cur.left:
                    if not cur.left.left and not cur.left.right:
                        sum = sum + cur.left.val
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)
        return sum

File: Golang/Sum of Left Leaves.go
// 1
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumOfLeftLeaves(root *TreeNode) int {
    if root == nil {
        return 0
    }
    if root.Left != nil && root.Left.Left == nil && root.Left.Right == nil {
        return root.Left.Val + sumOfLeftLeaves(root.Right)
    }
    return sumOfLeftLeaves(root.Left) + sumOfLeftLeaves(root.Right)
}

// 2
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumOfLeftLeaves(root *TreeNode) int {
    res := 0
    var dfs func(*TreeNode, bool)
    dfs = func(node *TreeNode, flag bool) {
        if node == nil {
            return
        }
        if node.Left == nil && node.Right == nil && flag {
            res += node.Val
        }
        dfs(node.Left, true)
        dfs(node.Right, false)
    }
    dfs(root, false)
    return res
}